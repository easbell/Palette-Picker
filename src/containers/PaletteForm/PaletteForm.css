.palette-form {
  width: 40%;
  margin: auto;
  padding: 20px;
  border-radius: 10px;
  background: #f4f4f4;
  color: #444;
  padding: 20px;
  box-shadow: 0 3px 5px rgba(0,0,0,0.20);
  display: flex;
  flex-direction: column;
}

.name-input, 
.project-input {
  padding: 10px;
  margin-bottom: 20px;
}

.pal-form-control {
  background: #fff;
  color: #757575;
  border: solid #bbbbbb 1px;
  padding: 10px 15px;
  font-size: 13px;
  font-weight: bold;
  letter-spacing: .02em;
  line-height: 17px;
  vertical-align: middle;
  cursor: pointer;
  text-decoration: none;
  border-radius: 4px;
}