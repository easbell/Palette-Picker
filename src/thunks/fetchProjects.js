import { isLoading, setProjects, hasErrored } from '../actions'

export const fetchProjects = () => {
  const url = process.env.REACT_APP_BACKEND_URL + 'api/v1/projects';
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const projects = await response.json()
      dispatch(isLoading(false))
      dispatch(setProjects(projects))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}