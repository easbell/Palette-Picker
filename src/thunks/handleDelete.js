import { isLoading, hasErrored } from '../actions'

export const handleDelete = (url, action, items, id) => {
  return async (dispatch) => {
    dispatch(isLoading(true))
    try {
      const response = await fetch(url, {
        method: 'DELETE',
        body: JSON.stringify({id: id}),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      const msg = await response.json()
      console.log(msg)
      dispatch(isLoading(false))
      dispatch(action(items))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}