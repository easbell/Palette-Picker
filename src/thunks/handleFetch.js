import { isLoading, hasErrored } from '../actions'

export const handleFetch = (url, action, options) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, options)
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      dispatch(isLoading(false))
      dispatch(action(data))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}

