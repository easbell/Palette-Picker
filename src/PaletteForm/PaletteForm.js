import React, { Component } from 'react';
import { addProject, setProjects, addPalette, setPalettes } from '../actions';
import { connect } from 'react-redux';
import { handleFetch } from '../thunks/handleFetch';
import DropdownButton from 'react-bootstrap/DropdownButton';
import Dropdown from 'react-bootstrap/Dropdown';
import './PaletteForm.css';

export class PaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      paletteName: '',
      projectName: '',
      newProject: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value })
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    await this.addProject()
    this.addPalette()
  }
  
  addProject = async () => {
    const { projectName } = this.state;
    const url = process.env.REACT_APP_BACKEND_URL + 'api/v1/projects/'
    const optionsObject = {
      method: 'POST',
      body: JSON.stringify({name: projectName}),
      headers: { 
        'Content-Type': 'application/json'
      }
    }
    await this.props.handleFetch(url, addProject, optionsObject);
    await this.props.handleFetch(url, setProjects);
  }
  
  addPalette = async (projectId) => {
    const { paletteName } = this.state
    const { projects, colors } = this.props
    const id = projectId || projects[projects.length -1].id
    const url = process.env.REACT_APP_BACKEND_URL + `api/v1/projects/${id}/palettes/`
    const allPalettesUrl = process.env.REACT_APP_BACKEND_URL + `api/v1/palettes/`    
    const optionsObject = {
      method: 'POST',
      body: JSON.stringify({
        palette_name: paletteName,
        color_1: colors[0],
        color_2: colors[1],
        color_3: colors[2],
        color_4: colors[3],
        color_5: colors[4]
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    await this.props.handleFetch(url, addPalette, optionsObject)
    this.props.savePalette(false);
    this.props.handleFetch(allPalettesUrl, setPalettes);
  }

  showProjects = () => {
    const { projects } = this.props;
    return projects.map(project => {
      return <Dropdown.Item onClick={() => {this.addPalette(project.id)}} key={project.id}>{project.name}</Dropdown.Item>
    })
  }

  newProject = () => {
    this.setState({ newProject: true })
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <h2>Save Palette</h2>
        <input 
          placeholder='Name this palette'
          name='paletteName'
          value={this.state.paletteName}
          onChange={this.handleChange}
        />
        <DropdownButton id="dropdown-basic-button" title="Saved Projects">
          <Dropdown.Item class="dropdown-item" onClick={this.newProject}>Add new project</Dropdown.Item>
          <div class="dropdown-divider"></div>
          {this.showProjects()}
        </DropdownButton>
        { this.state.newProject &&
          <input 
            placeholder='Name this project' 
            name='projectName'
            value={this.state.projectName}
            onChange={this.handleChange}
          />
        }
        <button type='submit'>Save palette & project</button>
      </form>
    )
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects
})

export const mapDispatchToProps = (dispatch) => ({
  handleFetch: (url, action, options) => dispatch(handleFetch(url, action, options))
})

export default connect(mapStateToProps, mapDispatchToProps)(PaletteForm)
