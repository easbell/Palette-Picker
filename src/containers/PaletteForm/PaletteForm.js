import React, { Component } from 'react';
import { addProject, setProjects, addPalette, setPalettes } from '../../actions';
import { connect } from 'react-redux';
import { handleFetch } from '../../thunks/handleFetch';
import DropdownButton from 'react-bootstrap/DropdownButton';
import Dropdown from 'react-bootstrap/Dropdown';
import './PaletteForm.css';
import cogoToast from 'cogo-toast';

export class PaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      paletteName: '',
      projectName: '',
      newProject: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value })
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const newName = this.state.projectName.toUpperCase()
    const projectNames = this.props.projects.map(project => project.name)
    if (projectNames.includes(newName)) {
      console.log(`${newName} already exists. Please choose another name for your project`)
    } else {
      await this.addProject(newName)
      this.addPalette()
    }
  }
  
  addProject = async (newName) => {
    const url = process.env.REACT_APP_BACKEND_URL + 'api/v1/projects/'
    const optionsObject = {
      method: 'POST',
      body: JSON.stringify({name: newName}),
      headers: { 
        'Content-Type': 'application/json'
      }
    }
    await this.props.handleFetch(url, addProject, optionsObject);
    await this.props.handleFetch(url, setProjects);
    cogoToast.success('Project was added.', {position: 'bottom-left'});
  }
  
  addPalette = async (projectId) => {
    const { paletteName } = this.state
    const { projects, colors } = this.props
    const id = projectId || projects[projects.length -1].id
    const url = process.env.REACT_APP_BACKEND_URL + `api/v1/projects/${id}/palettes/`
    const allPalettesUrl = process.env.REACT_APP_BACKEND_URL + `api/v1/palettes/`    
    const optionsObject = {
      method: 'POST',
      body: JSON.stringify({
        palette_name: paletteName,
        color_1: colors.color1.color,
        color_2: colors.color2.color,
        color_3: colors.color3.color,
        color_4: colors.color4.color,
        color_5: colors.color5.color
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    await this.props.handleFetch(url, addPalette, optionsObject)
    this.props.savePalette(false);
    this.props.handleFetch(allPalettesUrl, setPalettes);
    cogoToast.success('Palette was added.', {position: 'bottom-left'});
  }

  showProjects = () => {
    const { projects } = this.props;
    return projects.map(project => {
      return <Dropdown.Item onClick={() => {this.addPalette(project.id)}} key={project.id}>{project.name}</Dropdown.Item>
    })
  }

  newProject = () => {
    this.setState({ newProject: true })
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit} className='palette-form'>
        <input 
          placeholder='Name your palette'
          name='paletteName'
          className='name-input'
          value={this.state.paletteName}
          onChange={this.handleChange}
        />
        <DropdownButton id="dropdown-custom-1" title="Add to Project">
          <Dropdown.Item className="dropdown-item" onClick={this.newProject}>New project</Dropdown.Item>
          <div className="dropdown-divider"></div>
          {this.showProjects()}
        </DropdownButton>
        { this.state.newProject &&
          <input 
            placeholder='Name your new project' 
            name='projectName'
            value={this.state.projectName}
            onChange={this.handleChange}
            className='project-input'
          />
        }
        <button type='submit' className='pal-form-control'>Save Palette & Project</button>
      </form>
    )
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects
})

export const mapDispatchToProps = (dispatch) => ({
  handleFetch: (url, action, options) => dispatch(handleFetch(url, action, options))
})

export default connect(mapStateToProps, mapDispatchToProps)(PaletteForm)
