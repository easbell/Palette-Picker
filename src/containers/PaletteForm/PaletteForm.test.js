import React from 'react';
import { PaletteForm, mapStateToProps, mapDispatchToProps } from './PaletteForm';
import { handleFetch } from '../thunks/handleFetch';
import { shallow } from 'enzyme';

jest.mock('../thunks/handleFetch');

describe('PaletteForm', () => {
  let wrapper;
  let mockFn;
  let mockProjects;
  let mockColors;

  beforeEach(() => {
    mockFn = jest.fn();
    mockProjects = [{id: 1, name: 'PROJECT'}, {id: 2, name: 'project'}];
    mockColors = {
      color1: {color: 'red'},
      color2: {color: 'red'},
      color3: {color: 'red'},
      color4: {color: 'red'},
      color5: {color: 'red'},
    }

    wrapper = shallow(
      <PaletteForm 
        savePalette={mockFn}
        handleFetch={mockFn}
        projects={mockProjects}
        colors={mockColors}
      />
    )
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state upon change in form', () => {
    const mockEvent = { target: { name: 'paletteName', value: 'new palette'} }
    wrapper.find('.name-input').simulate('change', mockEvent);
    expect(wrapper.state('paletteName')).toBe('new palette');
  });

  it('should not invoke addProject if a project name already exist', async () => {
    const addProjectSpy = jest.spyOn(wrapper.instance(), 'addProject')
    wrapper.setState({projectName: 'project'})
    await wrapper.find('form').simulate('submit', {preventDefault: () => {}})
    expect(addProjectSpy).not.toHaveBeenCalled()
  })

  it('should invoke addProject and addPalette when handleSubmit is invoked', async () => {
    const addProjectSpy = jest.spyOn(wrapper.instance(), 'addProject')
    await wrapper.find('form').simulate('submit', {preventDefault: () => {}});
    expect(addProjectSpy).toHaveBeenCalled();
  });

  it('should dispatch handleFetch twice when addProject is invoked', async () => {
    wrapper.setState({ projectName: 'project' });
    await wrapper.instance().addProject();
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it('should dispatch savePalette and handleFetch when addPalette is invoked', async () => {
    await wrapper.instance().addPalette(3);
    expect(mockFn).toHaveBeenCalledTimes(3);
  });

  it('should update state when newProject is called', () => {
    wrapper.find('.dropdown-item').simulate('click');
    expect(wrapper.state('newProject')).toBe(true);
  });

  describe('mapStateToProps', () => {
    it('returns correct props', () => {
      const mockState = {
        projects: ['projects'],
        other: 'other'
      }
      const expectedProps = {
        projects: ['projects']
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expectedProps)
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch with handleFetch action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = handleFetch();
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.handleFetch();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});