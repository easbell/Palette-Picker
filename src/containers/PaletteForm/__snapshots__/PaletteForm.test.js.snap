// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteForm should match snapshot 1`] = `
<Bootstrap(Modal)
  className="modal palette-form"
  onHide={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="name-input"
      name="paletteName"
      onChange={[Function]}
      placeholder="Name this palette"
      value=""
    />
    <button
      className="pal-form-control disable"
      disabled={true}
      type="submit"
    >
      Save
    </button>
  </form>
</Bootstrap(Modal)>
`;
